plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/jobs/DataSourceMonitorJob.java

```diff
@@ -96,10 +96,12 @@ protected void doJob() {
     }
 
     protected void checkDataSourceAliveInWorkspace(DBPWorkspace workspace, long lastUserActivityTime) {
-        for (DBPProject project : workspace.getProjects()) {
+        List<DBPProject> projects = new ArrayList<>(workspace.getProjects());
```

Comment:
As far as I understood, another thread may be making changes to the array element while you are copying it. This could lead to an IndexOutOfBoundsException, uninitialized array values or even some sort of memory access exception, as System.arraycopy() is implemented in native code.